File: src/common/wasm/zig/partition.zig
//! WASM in-place partition (quickselect) kernels for all numeric types.
//!
//! Partitions an array so that a[kth] contains the value that would be
//! in position kth if the array were sorted. All elements before kth
//! are <= a[kth], all elements after are >= a[kth].
//!
//! Uses quickselect with median-of-three pivot selection.
//! Supported dtypes: f64, f32, i64, u64, i32, u32, i16, u16, i8, u8.

const sc = @import("sorting_common.zig");

/// Partition f64 array in-place. NaN values sort to end.
export fn partition_f64(a: [*]f64, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(f64, a, 0, N - 1, kth);
}

/// Partition f32 array in-place. NaN values sort to end.
export fn partition_f32(a: [*]f32, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(f32, a, 0, N - 1, kth);
}

/// Partition i64 array in-place.
export fn partition_i64(a: [*]i64, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(i64, a, 0, N - 1, kth);
}

/// Partition u64 array in-place.
export fn partition_u64(a: [*]u64, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(u64, a, 0, N - 1, kth);
}

/// Partition i32 array in-place.
export fn partition_i32(a: [*]i32, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(i32, a, 0, N - 1, kth);
}

/// Partition u32 array in-place.
export fn partition_u32(a: [*]u32, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(u32, a, 0, N - 1, kth);
}

/// Partition i16 array in-place.
export fn partition_i16(a: [*]i16, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(i16, a, 0, N - 1, kth);
}

/// Partition u16 array in-place.
export fn partition_u16(a: [*]u16, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(u16, a, 0, N - 1, kth);
}

/// Partition i8 array in-place.
export fn partition_i8(a: [*]i8, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(i8, a, 0, N - 1, kth);
}

/// Partition u8 array in-place.
export fn partition_u8(a: [*]u8, N: u32, kth: u32) void {
    if (N <= 1 or kth >= N) return;
    sc.quickselect(u8, a, 0, N - 1, kth);
}

// --- Batch slice partition ---

/// Partition numSlices contiguous f64 slices at kth position.
export fn partition_slices_f64(a: [*]f64, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(f64, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous f32 slices at kth position.
export fn partition_slices_f32(a: [*]f32, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(f32, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous i64 slices at kth position.
export fn partition_slices_i64(a: [*]i64, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(i64, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous u64 slices at kth position.
export fn partition_slices_u64(a: [*]u64, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(u64, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous i32 slices at kth position.
export fn partition_slices_i32(a: [*]i32, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(i32, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous u32 slices at kth position.
export fn partition_slices_u32(a: [*]u32, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(u32, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous i16 slices at kth position.
export fn partition_slices_i16(a: [*]i16, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(i16, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous u16 slices at kth position.
export fn partition_slices_u16(a: [*]u16, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(u16, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous i8 slices at kth position.
export fn partition_slices_i8(a: [*]i8, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(i8, a, sliceSize, numSlices, kth);
}

/// Partition numSlices contiguous u8 slices at kth position.
export fn partition_slices_u8(a: [*]u8, sliceSize: u32, numSlices: u32, kth: u32) void {
    sc.quickselectSlices(u8, a, sliceSize, numSlices, kth);
}

// --- Tests ---

test "partition_f64 basic" {
    const testing = @import("std").testing;
    var a = [_]f64{ 3.0, 1.0, 4.0, 1.5, 2.0 };
    partition_f64(&a, 5, 2);
    try testing.expectApproxEqAbs(a[2], 2.0, 1e-10);
    for (0..2) |i| try testing.expect(a[i] <= a[2]);
    for (3..5) |i| try testing.expect(a[i] >= a[2]);
}

test "partition_f64 NaN" {
    const testing = @import("std").testing;
    const nan = @as(f64, @bitCast(@as(u64, 0x7FF8000000000000)));
    var a = [_]f64{ nan, 2.0, 1.0, nan, 3.0 };
    partition_f64(&a, 5, 2);
    try testing.expectApproxEqAbs(a[2], 3.0, 1e-10);
}

test "partition_i32 basic" {
    const testing = @import("std").testing;
    var a = [_]i32{ 5, 3, -1, 4, 2 };
    partition_i32(&a, 5, 2);
    try testing.expectEqual(a[2], 3);
    for (0..2) |i| try testing.expect(a[i] <= a[2]);
    for (3..5) |i| try testing.expect(a[i] >= a[2]);
}

test "partition_u8 kth=0" {
    const testing = @import("std").testing;
    var a = [_]u8{ 50, 10, 40, 20, 30 };
    partition_u8(&a, 5, 0);
    try testing.expectEqual(a[0], 10);
}

test "partition_f64 single" {
    const testing = @import("std").testing;
    var a = [_]f64{42.0};
    partition_f64(&a, 1, 0);
    try testing.expectApproxEqAbs(a[0], 42.0, 1e-10);
}

File: src/common/wasm/partition.ts
/**
 * WASM-accelerated in-place partition (quickselect).
 *
 * Partitions a contiguous 1D buffer so that element at kth position
 * is in its final sorted position.
 * Returns null if WASM can't handle this case.
 */

import {
  partition_f64,
  partition_f32,
  partition_i64,
  partition_u64,
  partition_i32,
  partition_u32,
  partition_i16,
  partition_u16,
  partition_i8,
  partition_u8,
  partition_slices_f64,
  partition_slices_f32,
  partition_slices_i64,
  partition_slices_u64,
  partition_slices_i32,
  partition_slices_u32,
  partition_slices_i16,
  partition_slices_u16,
  partition_slices_i8,
  partition_slices_u8,
} from './bins/partition.wasm';
import {
  wasmMalloc,
  resetScratchAllocator,
  scratchCopyIn,
  getSharedMemory,
  f16InputToScratchF32,
} from './runtime';
import { ArrayStorage } from '../storage';
import { effectiveDType, hasFloat16, type DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type PartitionFn = (aPtr: number, N: number, kth: number) => void;
type SlicePartitionFn = (aPtr: number, sliceSize: number, numSlices: number, kth: number) => void;

const kernels: Partial<Record<DType, PartitionFn>> = {
  float64: partition_f64,
  float32: partition_f32,
  int64: partition_i64,
  uint64: partition_u64,
  int32: partition_i32,
  uint32: partition_u32,
  int16: partition_i16,
  uint16: partition_u16,
  int8: partition_i8,
  uint8: partition_u8,
  float16: partition_f32,
};

const sliceKernels: Partial<Record<DType, SlicePartitionFn>> = {
  float64: partition_slices_f64,
  float32: partition_slices_f32,
  int64: partition_slices_i64,
  uint64: partition_slices_u64,
  int32: partition_slices_i32,
  uint32: partition_slices_u32,
  int16: partition_slices_i16,
  uint16: partition_slices_u16,
  int8: partition_slices_i8,
  uint8: partition_slices_u8,
  float16: partition_slices_f32,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
  float16: Float32Array,
};

/**
 * WASM-accelerated partition of contiguous slices.
 * Uses batch kernel when slices are packed contiguously.
 *
 * Note: operates on pre-existing JS buffers, uses scratch.
 */
export function wasmPartitionSlices(
  resultData: TypedArray,
  sliceOffsets: Int32Array | number[],
  axisSize: number,
  outerSize: number,
  kth: number,
  dtype: DType
): boolean {
  if (axisSize < 2) return true;

  const sliceKernel = sliceKernels[dtype];
  const isF16 = dtype === 'float16';

  if (sliceKernel && sliceOffsets[0] === 0 && outerSize > 1 && sliceOffsets[1] === axisSize) {
    const Ctor = ctorMap[dtype];
    if (!Ctor) return false;

    wasmConfig.wasmCallCount++;
    resetScratchAllocator();

    const ptr = isF16
      ? f16InputToScratchF32(
          { data: resultData, isWasmBacked: false, wasmPtr: 0, offset: 0 },
          resultData.length
        )
      : scratchCopyIn(resultData as TypedArray);
    sliceKernel(ptr, axisSize, outerSize, kth);

    const mem = getSharedMemory();
    if (isF16) {
      const f16Out = new Uint16Array(resultData.length);
      const f32View = new Float32Array(mem.buffer, ptr, resultData.length);
      new Float16Array(f16Out.buffer, 0, resultData.length).set(f32View);
      new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
        new Uint8Array(f16Out.buffer, 0, f16Out.byteLength)
      );
    } else {
      new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
        new Uint8Array(mem.buffer, ptr, resultData.byteLength)
      );
    }
    return true;
  }

  // Fallback: per-slice calls
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return false;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const ptr = isF16
    ? f16InputToScratchF32(
        { data: resultData, isWasmBacked: false, wasmPtr: 0, offset: 0 },
        resultData.length
      )
    : scratchCopyIn(resultData as TypedArray);

  for (let i = 0; i < outerSize; i++) {
    kernel(ptr + sliceOffsets[i]! * bpe, axisSize, kth);
  }

  const mem = getSharedMemory();
  if (isF16) {
    const f32View = new Float32Array(mem.buffer, ptr, resultData.length);
    if (hasFloat16) {
      const f16Out = new Uint16Array(resultData.length);
      new Float16Array(f16Out.buffer, 0, resultData.length).set(f32View);
      new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
        new Uint8Array(f16Out.buffer, 0, f16Out.byteLength)
      );
    } else {
      // Polyfill: float16 arrays are backed by Float32Array, copy f32 values directly
      (resultData as Float32Array).set(f32View);
    }
  } else {
    new Uint8Array(resultData.buffer, resultData.byteOffset, resultData.byteLength).set(
      new Uint8Array(mem.buffer, ptr, resultData.byteLength)
    );
  }

  return true;
}

/**
 * WASM-accelerated partition of a contiguous 1D buffer.
 * Returns partitioned ArrayStorage or null if WASM can't handle it.
 */
export function wasmPartition(a: ArrayStorage, kth: number): ArrayStorage | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = effectiveDType(a.dtype);
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const outBytes = size * bpe;
  const isF16 = dtype === 'float16';

  // Partition is in-place: allocate output, copy input, partition in-place
  const outRegion = wasmMalloc(outBytes);
  if (!outRegion) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  const aOff = a.offset;
  if (isF16) {
    // Float16 partition: use the "sort floats as integers" trick.
    // Partition only rearranges elements — it doesn't modify values — so we can
    // transform f16 bit patterns into a form where i16 comparison gives the
    // correct f16 order, run partition_i16, and undo the transform.
    //
    // Transform: if sign bit is set (negative), XOR with 0x7FFF (flip exp+mantissa).
    // This makes signed i16 comparison match f16 ordering.
    // The transform is self-inverse: applying it again undoes it.
    const f16Bytes = size * 2;
    const f16Region = wasmMalloc(f16Bytes);
    if (!f16Region) {
      outRegion.release();
      return null;
    }
    const mem = getSharedMemory();

    // Copy f16 input into f16Region
    if (a.isWasmBacked) {
      new Uint8Array(mem.buffer, f16Region.ptr, f16Bytes).set(
        new Uint8Array(mem.buffer, a.wasmPtr + aOff * 2, f16Bytes)
      );
    } else {
      const aData = a.data.subarray(aOff, aOff + size) as TypedArray;
      new Uint8Array(mem.buffer, f16Region.ptr, f16Bytes).set(
        new Uint8Array(aData.buffer, aData.byteOffset, aData.byteLength)
      );
    }

    // Transform: make i16 comparison match f16 order
    const u16View = new Uint16Array(mem.buffer, f16Region.ptr, size);
    for (let j = 0; j < size; j++) {
      if (u16View[j]! & 0x8000) u16View[j]! ^= 0x7fff;
    }

    // Partition using i16 kernel
    partition_i16(f16Region.ptr, size, kth);

    // Undo transform
    for (let j = 0; j < size; j++) {
      if (u16View[j]! & 0x8000) u16View[j]! ^= 0x7fff;
    }

    outRegion.release();
    const F16Ctor = hasFloat16
      ? (Float16Array as unknown as new (
          buffer: ArrayBuffer,
          byteOffset: number,
          length: number
        ) => TypedArray)
      : (Uint16Array as unknown as new (
          buffer: ArrayBuffer,
          byteOffset: number,
          length: number
        ) => TypedArray);
    return ArrayStorage.fromWasmRegion(Array.from(a.shape), dtype, f16Region, size, F16Ctor);
  }

  // Copy input into output region, then partition in-place
  const mem = getSharedMemory();
  if (a.isWasmBacked) {
    new Uint8Array(mem.buffer, outRegion.ptr, outBytes).set(
      new Uint8Array(mem.buffer, a.wasmPtr + aOff * bpe, outBytes)
    );
  } else {
    const aData = a.data.subarray(aOff, aOff + size) as TypedArray;
    new Uint8Array(mem.buffer, outRegion.ptr, outBytes).set(
      new Uint8Array(aData.buffer, aData.byteOffset, aData.byteLength)
    );
  }

  kernel(outRegion.ptr, size, kth);

  return ArrayStorage.fromWasmRegion(
    Array.from(a.shape),
    dtype,
    outRegion,
    size,
    Ctor as unknown as new (buffer: ArrayBuffer, byteOffset: number, length: number) => TypedArray
  );
}
